packages/yarnpkg-core/sources/Project.ts

```diff
@@ -1179,15 +1186,19 @@ export class Project {
           if (typeof buildHash !== `undefined`)
             hash.update(buildHash);
 
-          traverse(resolution, new Set(seenPackages));
+          traverse(resolution);
```

Comment:
I think you need to `update` the hash with the `packageHashMap` you just stored (might as well make `traverse` return it). Otherwise the build hash of a package will be based on the build hash of its dependencies, but that's it (instead of being based on its whole dependency tree).

Reply:
For example if you have Foo -> Bar -> Baz and upgrade Baz, then Foo should be rebuilt.

Reply:
Yep, thanks, added